File: smash-ranks-android/app/src/main/java/com/garpr/android/adapters/BaseAdapter.java
package com.garpr.android.adapters;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<BaseAdapterViewHolder> {

    private final LayoutInflater mLayoutInflater;
    private final List<T> mItems;


    public BaseAdapter(@NonNull final Context context) {
        this(LayoutInflater.from(context));
    }

    public BaseAdapter(@NonNull final LayoutInflater layoutInflater) {
        mLayoutInflater = layoutInflater;
        mItems = new ArrayList<>();
    }

    public void add(@Nullable final T item) {
        if (item != null) {
            mItems.add(item);
            notifyDataSetChanged();
        }
    }

    public void add(@Nullable final List<T> items) {
        if (items != null && !items.isEmpty()) {
            for (final T item : items) {
                if (item != null) {
                    mItems.add(item);
                }
            }

            notifyDataSetChanged();
        }
    }

    public void clear() {
        if (!mItems.isEmpty()) {
            mItems.clear();
            notifyDataSetChanged();
        }
    }

    public Context getContext() {
        return mLayoutInflater.getContext();
    }

    public T getItem(final int position) {
        return mItems.get(position);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public List<T> getItems() {
        if (isEmpty()) {
            return Collections.emptyList();
        } else {
            return new ArrayList<>(mItems);
        }
    }

    @LayoutRes
    @Override
    public abstract int getItemViewType(final int position);

    public boolean isEmpty() {
        return mItems.isEmpty();
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onBindViewHolder(final BaseAdapterViewHolder holder, final int position) {
        final T item = mItems.get(position);
        ((BaseAdapterView) holder.itemView).setContent(item);
    }

    @Override
    public BaseAdapterViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        final View view = mLayoutInflater.inflate(viewType, parent, false);
        return new BaseAdapterViewHolder(view);
    }

    public void set(@Nullable final List<T> items) {
        mItems.clear();

        if (items != null && !items.isEmpty()) {
            mItems.addAll(items);
        }

        notifyDataSetChanged();
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/adapters/BaseAdapter.kt
package com.garpr.android.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView

abstract class BaseAdapter<T>(
        private val layoutInflater: LayoutInflater
) : RecyclerView.Adapter<BaseAdapterViewHolder>() {

    private val items = mutableListOf<T>()


    constructor(context: Context) : this(LayoutInflater.from(context))

    open fun clear() {
        if (!isEmpty) {
            items.clear()
            notifyDataSetChanged()
        }
    }

    val context: Context
        get() = layoutInflater.context

    fun getItem(position: Int) = items[position]

    override fun getItemCount() = items.size

    @LayoutRes
    abstract override fun getItemViewType(position: Int): Int

    val isEmpty: Boolean
        get() = items.isEmpty()

    override fun onBindViewHolder(holder: BaseAdapterViewHolder, position: Int) {
        val item = items[position]

        @Suppress("UNCHECKED_CAST")
        val view = holder.itemView as? BaseAdapterView<T> ?: throw ClassCastException(
                "itemView is an unknown type: ${holder.itemView}")

        view.setContent(item)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseAdapterViewHolder {
        val view = layoutInflater.inflate(viewType, parent, false)
        return BaseAdapterViewHolder(view)
    }

    open fun set(items: List<T>?) {
        this.items.clear()

        if (items?.isNotEmpty() == true) {
            this.items.addAll(items)
        }

        notifyDataSetChanged()
    }

}
